File: backend/app/elements/gui/widgets/plot_window/plot_window_element.py
"""The `plot_window` GUI widget element: display-only, no output port."""

from __future__ import annotations

from app.elements.base import DisplayWidget, Generation
from app.models.graph import GuiWidgetKind


class PlotWindowElement(DisplayWidget):
    """Ports, execute and authored_file are the display-widget contract; what
    is left is the one thing that makes this a chart rather than a picture."""

    widget_kind = GuiWidgetKind.PLOT_WINDOW
    config_fields = ("code", "code_prompt")

    def generation(self) -> Generation:
        """The transform reshapes data; it must not draw anything.

        Spelling that out matters: without it, models reliably reach for
        matplotlib, which is not installed in the sandbox and whose figures are
        not JSON-serializable anyway. The chart is drawn by the app
        (`PlotWidget.tsx`, a dependency-free SVG renderer).
        """
        return Generation(
            kind="code", prompt_field="code_prompt", target_field="code",
            contract=(
                'Must expose run(inputs: dict) -> dict, receiving {"value": <raw incoming data>} '
                'and returning {"value": <plot-ready data>}. Plot-ready data is a JSON-serializable '
                'list of points: either a list of numbers, or a list of {"label": str, "value": number} '
                'objects. The app renders these itself as an SVG bar/line chart -- do NOT draw anything '
                'and do NOT import plotting or third-party libraries (no matplotlib, plotly, pandas, '
                'numpy): the code runs in a sandbox with only the standard library available.'
            ),
            inputs=("value",), outputs=("value",),
            guard="Please describe the chart transform you need first.",
            success="✅ Transform generated!",
        )

File: frontend/src/elements/gui/widgets/plot_window/plotWindowElement.ts
import type { GuiWidgetElementDefinition } from '../../../types';
import PlotWindowEditor from './PlotWindowEditor';
import PlotWindowWidget from '../../../../components/gui/widgets/PlotWindowWidget';
import PlotWidget from '../../../../components/PlotWidget';

export const plotWindowElement: GuiWidgetElementDefinition = {
  widgetKind: 'plot_window',
  // Display-only, like an `output` node with write_mode="window": accepts data to
  // plot, no downstream port.
  authoredFile: (widget) => ({ extension: '.js', what: 'this transform' }),
  generation: {
    promptField: 'code_prompt',
    targetField: 'code',
    guard: 'Please describe the chart transform you need first.',
    success: '✅ Transform generated!',
    promptLabel: 'Prompt',
    promptPlaceholder: 'Describe the chart transform you need (axes, grouping, aggregation, filters, etc.)',
    // "Optional" on its own was a half-truth: without it the block is handed
    // whatever arrived, and unless that is already a list of points there is
    // nothing to draw.
    mono: true,
    bodyLabel: 'Plotting code — optional only if the incoming value is already points',
    bodyHeight: 100,
  },
  // What last arrived on this widget's input port, charted on the graph
  // canvas itself. GraphNodeComponent used to look for this widget kind by
  // name; it asks the element now.
  CanvasPreview: PlotWidget,
  ConfigEditor: PlotWindowEditor,
  RuntimeWidget: PlotWindowWidget,
};
